proposals/0436-recordwildcards-total.rst

```diff
@@ -0,0 +1,102 @@
+Proposal title
+==============
+
+.. author:: Brandon Chinn
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/436>`_.
+.. contents::
+
+``-XRecordWildCards`` does not give any warning or indication if you forget to use a field. This proposal adds syntax (optionally enabled by another extension) to give a warning if a field is unused.
+
+
+Motivation
+----------
+
+Some functions require all the fields of a record to be used, or at least considered, e.g. ``hash`` or ``toJSON``, or even something more security-facing. Currently, if one uses ``RecordWildCards`` and they add a field to the record later on, nothing will tell the developer to take a look at these functions.
+
+::
+
+ data Foo = Foo { a :: Int, b :: Bool }
+ 
+ instance ToJSON Foo where
+   toJSON Foo{..} =
+     object
+       [ "a" .= a
+       , "b" .= b
+       ]
+       
+ instance Hashable Foo where
+   hash Foo{..} = hash (a, b)
+ 
+ -- adding a 'c' field does not raise any indication of needing to look at `toJSON` or `hash`
+
+Existing workarounds are described in `this blog post <https://cs-syd.eu/posts/2021-09-10-undefined-trick>`_, all of which leave something to be desired:
+
+* Positional arguments break when fields are reordered, which matters in functions like `toJSON`
+* Using `undefined` is a bit of a hack, plus it fails `-Wunused-pattern-binds`. It also fails to explicitly say which field is unused (which isn't strictly necessary, but it's nice to have)
+
+The `safe-wild-cards <https://hackage.haskell.org/package/safe-wild-cards>`_ package provides a Template Haskell solution to this, but it would be nice to support this natively.
+
+One could even think of this as the dual of the `Or pattern proposal <https://github.com/ghc-proposals/ghc-proposals/pull/43>`_, to warn on unused terms in product types as opposed to unused terms in sum types.
+
+Proposed Change Specification
+-----------------------------
+
+When ``-XRecordWildCards`` is enabled, in addition to supporting ``Foo{..}`` in a pattern match, also support ``Foo{..!}``, which desugars to ``Foo{a = a, b = b}``.
```

Comment:
It would be nice to also support this in normal record pattern syntax, so I could write `Foo{a=anotherNameForA, b=anotherNameForB !}` and also be warned if another field is added. That might argue for a different syntax, however, since it looks stranger in non-wildcard case.

Reply:
```
Prelude> data Foo = Foo { foo :: Int, bar :: Char} deriving Show
Prelude> :set -Wall
Prelude> Foo { foo = 1 }

<interactive>:3:1: warning: [-Wmissing-fields]
    • Fields of ‘Foo’ not initialised: bar
    • In the expression: Foo {foo = 1}
      In an equation for ‘it’: it = Foo {foo = 1}
```

Reply:
I strongly agree with this sentiment.

By the way, for comparison, in Ocaml, record matching must be exhaustive by default. The following is an error:

```ocaml
match { a; b; c } with
| { a; b } -> …
```

If you want to drop some fields, you need special syntax

```ocaml
match { a; b; c } with
| { a; b; .. } -> …
```

(here, contrary to Haskell's syntax, the `..` do not bind anything. It just means “ignore the remaining fields”)

Reply:
> for comparison, in Ocaml, record matching must be exhaustive by default

Same in Rust. I prefer Rust's version to Haskell's (if you want to use many fields, just write `x.a` `x.b` `x.c` and then it will be clearer where the values are coming from), but having `{..!}` is definitely better than nothing.